src/sentry/utils/http.py

```diff
@@ -222,3 +224,37 @@ def origin_from_request(request):
     if rv in ('', 'null'):
         rv = origin_from_url(request.META.get('HTTP_REFERER'))
     return rv
+
+
+def heuristic_decode(data, possible_content_Type=None):
+    """
+    Attempt to decode a HTTP body by trying JSON and Form URL decoders,
+    returning the decoded body (if decoding was successful) and the interred
+    content type.
+    """
+    inferred_content_type = possible_content_Type
+
+    form_encoded_parser = partial(
+        parse_qs,
+        strict_parsing=True,
+        keep_blank_values=True,
+    )
+
+    decoders = [
+        ('application/x-www-form-urlencoded', form_encoded_parser),
+        ('application/json', json.loads),
+    ]
```

Comment:
Defining these in the function body (rather than at the module level) seems a bit odd to me, but I suppose if you're going to be sorting in-place below it makes sense.